nbclassic 1.3.0 fails to build offline

## Description

The 1.3.0 nbclassic package can no longer be built offline.

## Reproduce

```console
$ tar -xf dist/nbclassic-1.3.0.tar.gz
$ cd nbclassic-1.3.0
$ uv venv
Using CPython 3.13.3 interpreter at: /usr/lib/python-exec/python3.13/python3
Creating virtual environment at: .venv
Activate with: source .venv/bin/activate
$ uv pip install build hatchling setuptools babel jupyter-server hatch-jupyter-builder
Resolved 67 packages in 32ms
░░░░░░░░░░░░░░░░░░░░ [0/67] Installing wheels...                                                                                       warning: Failed to hardlink files; falling back to full copy. This may lead to degraded performance.
         If the cache and target directories are on different filesystems, hardlinking may not be supported.
         If this is intentional, set `export UV_LINK_MODE=copy` or use `--link-mode=copy` to suppress this warning.
Installed 67 packages in 100ms
 + anyio==4.9.0
 + argon2-cffi==23.1.0
 + argon2-cffi-bindings==21.2.0
 + arrow==1.3.0
 + attrs==25.3.0
 + babel==2.17.0
 + beautifulsoup4==4.13.4
 + bleach==6.2.0
 + build==1.2.2.post1
 + cffi==1.17.1
 + defusedxml==0.7.1
 + fastjsonschema==2.21.1
 + fqdn==1.5.1
 + hatch-jupyter-builder==0.9.1
 + hatchling==1.27.0
 + idna==3.10
 + isoduration==20.11.0
 + jinja2==3.1.6
 + jsonpointer==3.0.0
 + jsonschema==4.23.0
 + jsonschema-specifications==2025.4.1
 + jupyter-client==8.6.3
 + jupyter-core==5.7.2
 + jupyter-events==0.12.0
 + jupyter-server==2.15.0
 + jupyter-server-terminals==0.5.3
 + jupyterlab-pygments==0.3.0
 + markupsafe==3.0.2
 + mistune==3.1.3
 + nbclient==0.10.2
 + nbconvert==7.16.6
 + nbformat==5.10.4
 + overrides==7.7.0
 + packaging==25.0
 + pandocfilters==1.5.1
 + pathspec==0.12.1
 + platformdirs==4.3.7
 + pluggy==1.5.0
 + prometheus-client==0.21.1
 + ptyprocess==0.7.0
 + pycparser==2.22
 + pygments==2.19.1
 + pyproject-hooks==1.2.0
 + python-dateutil==2.9.0.post0
 + python-json-logger==3.3.0
 + pyyaml==6.0.2
 + pyzmq==26.4.0
 + referencing==0.36.2
 + rfc3339-validator==0.1.4
 + rfc3986-validator==0.1.1
 + rpds-py==0.24.0
 + send2trash==1.8.3
 + setuptools==79.0.1
 + six==1.17.0
 + sniffio==1.3.1
 + soupsieve==2.7
 + terminado==0.18.1
 + tinycss2==1.4.0
 + tornado==6.4.2
 + traitlets==5.14.3
 + trove-classifiers==2025.4.11.15
 + types-python-dateutil==2.9.0.20241206
 + typing-extensions==4.13.2
 + uri-template==1.3.0
 + webcolors==24.11.1
 + webencodings==0.5.1
 + websocket-client==1.8.0
$ sudo unshare -n sudo -u $USER .venv/bin/python -m build -nw
* Getting build dependencies for wheel...
* Building wheel...
INFO:hatch_jupyter_builder.utils:Running jupyter-builder
INFO:hatch_jupyter_builder.utils:Building with hatch_jupyter_builder.npm_builder
INFO:hatch_jupyter_builder.utils:With kwargs: {}
WARNING:hatch_jupyter_builder.utils:yarn not found, ignoring yarn.lock file
INFO:hatch_jupyter_builder.utils:Installing build dependencies with npm.  This may take a while...
INFO:hatch_jupyter_builder.utils:> /usr/sbin/npm install
npm warn ERESOLVE overriding peer dependency
npm warn While resolving: @lumino/coreutils@1.12.1
npm warn Found: crypto@undefined
npm warn node_modules/crypto
npm warn
npm warn Could not resolve dependency:
npm warn peer crypto@"1.0.1" from @lumino/coreutils@1.12.1
npm warn node_modules/@lumino/coreutils
npm warn   @lumino/coreutils@"^1.11.0" from @jupyterlab/apputils@3.6.8
npm warn   node_modules/@jupyterlab/apputils
npm warn   12 more (@jupyterlab/coreutils, @jupyterlab/nbformat, ...)
npm error code 1
npm error The git reference could not be found
npm error command git --no-replace-objects checkout ~1.6.0
npm error error: pathspec '~1.6.0' did not match any file(s) known to git
npm error A complete log of this run can be found in: /home/mgorny/.npm/_logs/2025-04-25T04_30_51_837Z-debug-0.log
Traceback (most recent call last):
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/pyproject_hooks/_in_process/_in_process.py", line 389, in <module>
    main()
    ~~~~^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/pyproject_hooks/_in_process/_in_process.py", line 373, in main
    json_out["return_val"] = hook(**hook_input["kwargs"])
                             ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/pyproject_hooks/_in_process/_in_process.py", line 280, in build_wheel
    return _build_backend().build_wheel(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        wheel_directory, config_settings, metadata_directory
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatchling/build.py", line 58, in build_wheel
    return os.path.basename(next(builder.build(directory=wheel_directory, versions=['standard'])))
                            ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatchling/builders/plugin/interface.py", line 147, in build
    build_hook.initialize(version, build_data)
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatch_jupyter_builder/plugin.py", line 94, in initialize
    raise e
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatch_jupyter_builder/plugin.py", line 89, in initialize
    build_func(self.target_name, version, **build_kwargs)
    ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatch_jupyter_builder/utils.py", line 115, in npm_builder
    run([*npm_cmd, "install"], cwd=str(abs_path))
    ~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatch_jupyter_builder/utils.py", line 231, in run
    return subprocess.check_call(cmd, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/subprocess.py", line 419, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/sbin/npm', 'install']' returned non-zero exit status 1.

ERROR Backend subprocess exited when trying to invoke build_wheel
```

If I install yarn, it fails more explicitly:

```console
$ sudo unshare -n sudo -u $USER .venv/bin/python -m build -nw
* Getting build dependencies for wheel...
* Building wheel...
INFO:hatch_jupyter_builder.utils:Running jupyter-builder
INFO:hatch_jupyter_builder.utils:Building with hatch_jupyter_builder.npm_builder
INFO:hatch_jupyter_builder.utils:With kwargs: {}
INFO:hatch_jupyter_builder.utils:Installing build dependencies with npm.  This may take a while...
INFO:hatch_jupyter_builder.utils:> /usr/sbin/yarn install
yarn install v1.22.22
warning You don't appear to have an internet connection. Try the --offline flag to use the cache for registry queries.
info There appears to be trouble with your network connection. Retrying...
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
error Error: getaddrinfo ENOTFOUND registry.yarnpkg.com
    at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:dns:120:26)
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
Traceback (most recent call last):
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/pyproject_hooks/_in_process/_in_process.py", line 389, in <module>
    main()
    ~~~~^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/pyproject_hooks/_in_process/_in_process.py", line 373, in main
    json_out["return_val"] = hook(**hook_input["kwargs"])
                             ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/pyproject_hooks/_in_process/_in_process.py", line 280, in build_wheel
    return _build_backend().build_wheel(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        wheel_directory, config_settings, metadata_directory
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatchling/build.py", line 58, in build_wheel
    return os.path.basename(next(builder.build(directory=wheel_directory, versions=['standard'])))
                            ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatchling/builders/plugin/interface.py", line 147, in build
    build_hook.initialize(version, build_data)
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatch_jupyter_builder/plugin.py", line 94, in initialize
    raise e
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatch_jupyter_builder/plugin.py", line 89, in initialize
    build_func(self.target_name, version, **build_kwargs)
    ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatch_jupyter_builder/utils.py", line 115, in npm_builder
    run([*npm_cmd, "install"], cwd=str(abs_path))
    ~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/nbclassic-1.3.0/.venv/lib/python3.13/site-packages/hatch_jupyter_builder/utils.py", line 231, in run
    return subprocess.check_call(cmd, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/subprocess.py", line 419, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/sbin/yarn', 'install']' returned non-zero exit status 1.

ERROR Backend subprocess exited when trying to invoke build_wheel
```

## Expected behavior

Build working, like in 1.2.0.

## Context

- Operating System and version: Gentoo Linux amd64
- Browser and version: n/a
- Jupyter NbClassic version: 1.3.0

I'm trying to update the nbclassic package in Gentoo. For downstreams, is it absolute crucial that [builds work offline](https://packaging.python.org/en/latest/discussions/downstream-packaging/#do-not-use-the-internet-during-the-build-process). Besides, implicitly fetching dependencies from Internet poses a privacy (and potentially a security) risk.

--- a/.github/workflows/check-release.yml
+++ b/.github/workflows/check-release.yml
@@ -49,7 +49,7 @@ jobs:
             ${{ runner.os }}-linkcheck-
       - name: Upgrade packaging dependencies
         run: |
-          pip install --upgrade pip setuptools wheel --user
+          pip install --upgrade pip wheel --user
       - name: Install Dependencies
         run: |
           pip install -e .

--- a/.github/workflows/docs.yml
+++ b/.github/workflows/docs.yml
@@ -27,7 +27,7 @@ jobs:
         architecture: 'x64'
     - name: Upgrade packaging dependencies
       run: |
-        pip install --upgrade pip setuptools wheel
+        pip install --upgrade pip wheel
     - name: Get pip cache dir
       id: pip-cache
       run: |

--- a/.github/workflows/js.yml
+++ b/.github/workflows/js.yml
@@ -73,7 +73,6 @@ jobs:
       - name: Install dependencies
         run: |
           pip install --upgrade pip
-          pip install --upgrade setuptools wheel
           npm run build
           npm install -g casperjs@1.1.4 phantomjs-prebuilt@2.1.16
           pip install .[test]

--- a/.github/workflows/pythonpackage.yml
+++ b/.github/workflows/pythonpackage.yml
@@ -133,3 +133,24 @@ jobs:
         run: |
           cd sdist/test
           pytest -vv || pytest -vv --lf
+  test_offline_build:
+    runs-on: ubuntu-latest
+    needs: [make_sdist]
+    name: Test building offline
+    timeout-minutes: 20
+    steps:
+      - name: Base Setup
+        uses: jupyterlab/maintainer-tools/.github/actions/base-setup@v1
+      - name: Download sdist
+        uses: actions/download-artifact@v4
+      - name: Install From SDist
+        run: |
+          set -ex
+          cd sdist
+          mkdir test
+          tar --strip-components=1 -zxvf *.tar.gz -C ./test
+      - name: Test offline build with unshare
+        run: |
+          # Use unshare to create a network namespace that blocks network access
+          cd test
+          sudo unshare --net bash -c "python -m build --no-isolation --wheel"

--- a/.github/workflows/pythonpackage.yml
+++ b/.github/workflows/pythonpackage.yml
@@ -153,4 +153,6 @@ jobs:
         run: |
           # Use unshare to create a network namespace that blocks network access
           cd sdist/test
-          sudo unshare --net bash -c "python -m build --no-isolation --wheel"
+          python -m venv .venv
+          .venv/bin/python -m pip install build babel jupyter-server hatch-jupyter-builder
+          sudo unshare --net bash -c ".venv/bin/python -m build --no-isolation --wheel"

--- a/.github/workflows/pythonpackage.yml
+++ b/.github/workflows/pythonpackage.yml
@@ -152,5 +152,5 @@ jobs:
       - name: Test offline build with unshare
         run: |
           # Use unshare to create a network namespace that blocks network access
-          cd test
+          cd sdist/test
           sudo unshare --net bash -c "python -m build --no-isolation --wheel"

--- a/CONTRIBUTING.rst
+++ b/CONTRIBUTING.rst
@@ -49,7 +49,7 @@ Installing the Jupyter NbClassic
 Once you have installed the dependencies mentioned above, use the following
 steps::
 
-    pip install --upgrade setuptools pip
+    pip install --upgrade pip
     git clone https://github.com/jupyter/nbclassic
     cd nbclassic
     pip install -e .

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -5,7 +5,6 @@ requires = [
   "babel",
   "hatchling",
   "jupyter-server>=1.17",
-  "setuptools",
 ]
 
 [project]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -163,6 +163,11 @@ dependencies = [
   "hatch-jupyter-builder>=0.9.1",
 ]
 build-function = "hatch_jupyter_builder.npm_builder"
+# Allow offline builds of sdist, prevents running yarn install again
+skip-if-exists = ["nbclassic/static/components/MathJax/MathJax.js"]
+
+[tool.hatch.build.hooks.jupyter-builder.build-kwargs]
+npm = ["yarn"]
 
 [tool.pytest.ini_options]
 addopts = "--doctest-modules"

--- a/tools/install_pydeps.py
+++ b/tools/install_pydeps.py
@@ -29,7 +29,7 @@ def attempt(arg_list, max_attempts=1, name=''):
 
 def run():
     steps = {
-        'step1': """python -m pip install -U pip setuptools wheel""".split(' '),
+        'step1': """python -m pip install -U pip""".split(' '),
         'step2': """pip install pytest-playwright""".split(' '),
         'step3': """playwright install""".split(' '),
         'step4': """pip install .[test]""".split(' '),

